tracing/middleware.go

```diff
@@ -0,0 +1,114 @@
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+	"time"
+
+	"github.com/bytedance/gopkg/cloud/metainfo"
+	"github.com/cloudwego/kitex/pkg/endpoint"
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/kitex-contrib/obs-opentelemetry/tracing/internal"
+	"go.opentelemetry.io/otel/baggage"
+	"go.opentelemetry.io/otel/sdk/trace"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+// ClientMiddleware inject span context into req meta
+func ClientMiddleware(cfg *config) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, req, resp interface{}) (err error) {
+			span := oteltrace.SpanFromContext(ctx)
+			if !span.IsRecording() {
+				return next(ctx, req, resp)
+			}
+
+			// inject client service resource attributes (canonical service) to baggage
+			readOnlySpan := span.(trace.ReadOnlySpan)
+			bags, err := injectCanonicalServiceToBaggage(readOnlySpan.Resource().Attributes())
+			if err != nil {
+				return err
+			}
+			ctx = baggage.ContextWithBaggage(ctx, bags)
+
+			// inject to meta
+			md := metainfo.GetAllValues(ctx)
+			if md == nil {
+				md = make(map[string]string)
+			}
+			Inject(ctx, cfg, md)
+			for k, v := range md {
+				ctx = metainfo.WithValue(ctx, k, v)
+			}
+
+			if err = next(ctx, req, resp); err != nil {
+				RecordErrorSpan(span, err, cfg.withStackTrace)
+			}
+			return err
+		}
+	}
+}
+
+// ServerMiddleware extract req meta into span context
+func ServerMiddleware(cfg *config) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, req, resp interface{}) (err error) {
+			tc := internal.TraceCarrierFromContext(ctx)
+			if tc == nil {
+				klog.CtxWarnf(ctx, "TraceCarrier not found in context")
+				return next(ctx, req, resp)
+			}
+
+			// get tracer from carrier
+			sTracer := tc.Tracer()
+
+			var spanName string
+			if cfg.spanNameFormatter != nil {
+				spanName = cfg.spanNameFormatter(ctx)
+			}
+
+			ri := rpcinfo.GetRPCInfo(ctx)
+			opts := []oteltrace.SpanStartOption{
+				oteltrace.WithTimestamp(getStartTimeOrDefault(ri, time.Now())),
+				oteltrace.WithSpanKind(oteltrace.SpanKindServer),
+			}
+
+			md := metainfo.GetAllValues(ctx)
+			bags, spanCtx := Extract(ctx, cfg, md)
+			ctx = baggage.ContextWithBaggage(ctx, bags)
+
+			ctx, span := sTracer.Start(oteltrace.ContextWithRemoteSpanContext(ctx, spanCtx), spanName, opts...)
+
+			// peer service resource attrs
+			attrs := peerServiceAttributesFromBaggage(bags)
+			span.SetAttributes(attrs...)
+
+			// set span and attrs into tracer carrier for serverTracer finish
+			tc.SetSpan(span)
+
+			// reset service baggage
+			bags = resetPeerServiceBaggageMember(bags)
+			ctx = baggage.ContextWithBaggage(ctx, bags)
+
+			if err = next(ctx, req, resp); err != nil {
+				RecordErrorSpan(span, err, cfg.withStackTrace)
```

Comment:
rpcStats.Error() will return a not nil err if next() return a not nil err. So the err will be recorded twice(`tracer_server.go:112`).

Reply:
> rpcStats.Error() will return a not nil err if next() return a not nil err. So the err will be recorded twice(`tracer_server.go:112`).

Yes, it seems that there is no need to record errors here.